### PR TITLE
Autosetters with Subpkgs

### PR DESCRIPTION
**Pull request by phanimarupaka**
@mortent 

This PR is to set the auto-setters recursively in all the subpackages when the package is fetched using `kpt pkg get`

### CODE REVIEW

File: internal/util/setters/setters.go [hunk 1]
```diff
@@ -96,48 +95,69 @@ func PerformSetters(path string) error {
...
 
-		if c == "core.project" && (setter.Count > 0 || setter2.Count > 0) {
-			// set the projectNumber if we set the projectID
-			projectID = v
-		}
+	return nil
+}
+
+var GetProjectNumberFromProjectID = func(projectID string) (string, error) {
+	gcloudCmd := exec.Command("gcloud",
```
Comment (mortent) [inline internal/util/setters/setters.go:113]: How do we handle situations where gcloud is not available?

File: internal/util/setters/setters.go [hunk 2]
```diff
@@ -96,48 +95,69 @@ func PerformSetters(path string) error {
...
+		return "", errors.Wrapf(err, "failed to get project number for %s, please verify gcloud "+
+			"credentials are valid and try again", projectID)
 	}
+	return strings.TrimSpace(string(b)), nil
+}
 
-	if projectID != "" {
+// DefExists returns true if the setterName exists in Kptfile definitions
+func DefExists(resourcePath, setterName string) bool {
+	if err := openapi.AddSchemaFromFile(filepath.Join(resourcePath, "Kptfile")); err != nil {
```
Comment (mortent) [inline internal/util/setters/setters.go:125]: The Kptfile name must be defined as a constant somewhere?

File: internal/util/setters/setters.go [hunk 3]
```diff
@@ -158,43 +178,68 @@ func PerformSetters(path string) error {
...
 }
 
-func GetProjectNumberFromProjectID(projectID string) (string, error) {
-	gcloudCmd := exec.Command("gcloud",
-		"projects", "describe", projectID, "--format", "value(projectNumber)")
-	b, err := gcloudCmd.Output()
+// SetV2AutoSetter sets the input auto setter recursively in all the sub-packages of root
+// Sets GcloudProjectNumber as well, if input setter is GcloudProject
+func SetV2AutoSetter(name, value, root string) error {
+	openAPIFileName := "Kptfile"
```
Comment (mortent) [inline internal/util/setters/setters.go:189]: Same as above. I assume we have this as a constant somewhere

File: internal/util/update/resource-merge.go [hunk 4]
```diff
@@ -224,7 +224,7 @@ func getSubDirsAndNonKrmFiles(root string) (sets.String, sets.String, error) {
 			}
 			return nil
 		}
-		match, err := r.ShouldSkipFile(info)
+		match, err := r.ShouldSkipFile(info.Name(), &kio.IgnoreFilesMatcher{})
```
Comment (mortent) [inline internal/util/update/resource-merge.go:227]: I don't think this is the right way to fix this. I have a fix for this in https://github.com/GoogleContainerTools/kpt/pull/996.

File: internal/util/setters/setters.go (hunk 1 above)
Comment (phanimarupaka) [reply internal/util/setters/setters.go:113]: This method is execute iff gcloud.core.project is set which means the gcloud exists. If the gcloud is not installed we continue and not error out. Refer to line 90 comment.

File: internal/util/setters/setters.go (hunk 2 above)
Comment (phanimarupaka) [reply internal/util/setters/setters.go]: Done.

File: internal/util/setters/setters.go (hunk 3 above)
Comment (phanimarupaka) [reply internal/util/setters/setters.go]: Done.

File: internal/util/update/resource-merge.go (hunk 4 above)
Comment (phanimarupaka) [reply internal/util/update/resource-merge.go]: Rebased.

File: internal/util/setters/setters.go (hunk 1 above)
Comment (mortent) [reply internal/util/setters/setters.go:114]: 👍

**Pull request merged**